### PR TITLE
use findAddressCandidates consistently

### LINKED ISSUE #107: find > findAddressCandidates
> The `find` operation has been deprecated. Please discontinue using it in your application, and use the `findAddressCandidates` operation instead.

reference: [../api-reference/geocoding-find-address-candidates.htm](https://developers.arcgis.com/rest/geocode/api-reference/geocoding-find-address-candidates.htm)

### PR DESCRIPTION
**Pull request by jgravois**
resolves #107

this change makes consistent use of the findAddressCandidates operation of the World Geocoding Service.  this makes it unnecessary to differentiate and maintain simple() and addresses() methods for geocoding.

wrote out a few more samples in the markdown doc to demonstrate how to pass additional parameters and make requests to custom services.

because the response signature of findAddressCandidates differs from find, it seemed like a good time to group some additional cleanup and just call the whole thing a breaking change.

### CODE REVIEW

File: docs/Geocoding.md [hunk 1]
```diff
@@ -12,21 +12,42 @@ documented here are based on that API. For the most current and official informa
...
     });
 ```
 
 `client.geocode` is a shortcut for calling `client.geocode.simple`.
-
+`text` is mapped automatically as the geocoder parameter `singleLine`.
+
+## Passing additional parameters
+
+Any parameter supported by the [`findAddressCandidates'](https://developers.arcgis.com/rest/geocode/api-reference/geocoding-find-address-candidates.htm) operation of the World Geocoding Service can be passed through.
```
Comment (nixta) [inline docs/Geocoding.md]: Start and and quotes do not match.

File: lib/geocode.js [hunk 2]
```diff
@@ -53,11 +52,30 @@ function _handleArcGISResponse (error, response, body, callback) {
  * geoservice.geocode({ text: "920 SW 3rd Ave, Portland, OR 97204" }, callback);
 */
 function geocode (parameters, callback) {
-  parameters.f = parameters.f || "json";
```
Comment (nixta) [inline lib/geocode.js]: Is this functionally different? Or are you just going for clarity?

File: test/geocode-test.js [hunk 3]
```diff
@@ -7,24 +7,28 @@ var geocode = require('../lib/geocode');
...
+      geocode.geocode({ 
+        text: "920 SW 3rd Ave, Portland, OR 97204"
+      }, this.callback);
     },
     'It should return the correct latitude and longitude': function (err, data) {
       assert.equal(err, null);
-      assert.equal(data.locations.length, 1);
-      assert.equal(data.locations[0].feature.geometry.x.toPrecision(5), (-122.67633658436517).toPrecision(5));
-      assert.equal(data.locations[0].feature.geometry.y.toPrecision(5), (45.5167324388521).toPrecision(5));
+      assert.equal(data.candidates.length, 10);
```
Comment (nixta) [inline test/geocode-test.js]: Does this need to be `=10` or `>0`? Not sure if it's a default service setting, but I guess the question is: Are you testing that you get a result or that the service is maxing out the results returned?

File: lib/geocode.js (hunk 2 above)
Comment (jgravois) [inline lib/geocode.js]: > are you just going for clarity?

i didn't write that line, and the World Geocoding Service doesn't even support `html` output, but my guess is that this line was written to ensure that developers *can* ask for 'html' from their own geocoding services if they want to.

does that make sense? am i even answering your question?

File: test/geocode-test.js (hunk 3 above)
Comment (jgravois) [inline test/geocode-test.js]: `> 0` would be sufficient. my impression is that the test only verifies that the request actually generates results.

if i was more ambitious, i'd rewrite the tests with a mock server and skip verifying a lot of the response altogether.

thanks for all this **good** feedback!

**Comment by jgravois** [comment]
you are all good with getting rid of `addresses()` entirely?
you see any other low hanging / breaking change type fruit?
what would you say this PR means in the way of SemVer?

File: lib/geocode.js (hunk 2 above)
Comment (nixta) [inline lib/geocode.js]: Oh, it was that change from this to the more explicit `if (!parameters.f) {}` that I was wondering about. I get what it's doing and why, just wondering why the change.

Comment (jgravois) [inline lib/geocode.js]: wow. good catch.  

its actually just an [existing inconsistency](https://github.com/Esri/geoservices-js/blob/c1f5c51dc669c6ec232e750eb36f10871b0c93dd/lib/geocode.js#L88-L90) that i copied over.

i'm 8 days foggy, but willing to give myself the benefit of the doubt and say i was 'going for clarity'

**Comment by jgravois** [comment]
> I also recognize that's not the way the agile kids work these days.

if we had any other updates to push, i'd be all for doing some sort of intermediate deprecation, but since we don't...

another option i guess is to just leave the public method in there and quit mentioning it in the docs. making the lib more approachable for folks using it for the first time is really my primary goal.

File: lib/geocode.js (hunk 2 above)
Comment (nixta) [inline lib/geocode.js]: I'm good either way. Having consistency is better than not. This is one of those where everything is a grey area.

**Comment by nixta** [comment]
I like that. Add a comment in the address() function explaining to use the new method and this one might go away at the next major version bump?

**Comment by jgravois** [comment]
:1st_place_medal: remind me to make you start reviewing more of my pull requests

**Comment by jgravois** [comment]
the geocoding team has confirmed that their upcoming release doesn't include any known breaking changes in request/response signatures.

at this point, i don't see a way to simplify this API and start using `/findAddressCandidates` exclusively *without* making a breaking change, so i say we just rip off the proverbial bandaid.

**Comment by nixta** [comment]
Ripping the bandaid? Removing the function, aligning this with the geocoded, and bumping the version?

**Comment by jgravois** [comment]
exactly.

**Comment by nixta** [comment]
Ok. Let's do it.

**Comment by JerrySievert** [comment]
do it! just bump up a full revision, please :)

**Comment by jgravois** [comment]
i'm probably overthinking this, but I decided to take the middle ground and leave the text > **s**ingleLine mapping in the library and just stop documenting it.
